SHS-6317: Add CodeCeption support in DDEV

nJim opened the pull request:
# Summary
- Updates the DDEV config and services to use containers that are capable of running CodeCeption on a machine using the ARM architecture. This will unblock Mac M1 users who were not able to run tests locally.
- Migrates the setup steps from Lando to DDEV. Defines a setup script to automate these steps and make them repeatable.
- Updates documentation for CodeCeption testing on DDEV.

## Backstory
- Automated testing has many benefits. This is a useful tool but for many developers this is only running on GitHub within the CI workflow. Failing at this point is too late in the development cycle with engineers needed to push a change, wait for the test to fail, and then try again. This ticket will unblock engineers who want to run CodeCeption tests on their localdev environment in a way that matches the BLT-based workflow in GitHub.

## Need Review By (Date)
Not set.

## Urgency
Low

## Steps to Test
1. Review DDEV documentation for accuracy. 
2. If already running DDEV, delete the local containers including snapshots: `ddev delete --omit-snapshot`.
3. Start DDEV to rebuild the local environment: `ddev start`.
4. Setup the testing environment to run CodeCeption tests: `ddev setup-tests`.
5. Run the tests using blt and verify that all tests pass: `blt codeception --group=install --suite=acceptance`
6. Notice that this process does modify `docroot/sites/default/settings.php`. This change should not be committed. There may be other ways to to avoid this artifact, but we are trying to stay away from changing anything outside of the .ddev folder so that these changes do not impact engineers who are not using ddev.

# Review Tasks

## Backend / Functional Validation
### Code
- [ ] Are the naming conventions following our standards?
- [ ] Are PHP functions and variables in `snake_case` and not `camelCase`?
- [ ] Does Drupal code follow [Drupal Coding Standards](https://www.drupal.org/docs/develop/standards/php/php-coding-standards)?
- [ ] Does the code have sufficient inline comments?
- [ ] Is there anything in this code that would be hidden or hard to discover through the UI?
- [ ] Are there any [code smells](https://blog.codinghorror.com/code-smells/)?
- [ ] Are tests provided?

### Code security
- [ ] Is all [user input properly sanitized when rendered](https://www.drupal.org/docs/8/security/drupal-8-sanitizing-output)?
- [ ] Any obvious [security flaws or new areas for attack](https://www.drupal.org/docs/8/security)?

### General
- [ ] Is there anything included in this PR that is not related to the problem it is trying to solve?
- [ ] Is the approach to the problem appropriate?

nJim committed:
feat(SHS-6316): migration from lando to ddev for localdev

nJim committed:
feat(SHS-6316): prevent ddev updates to settings.php

nJim committed:
feat(SHS-6316): update ddev docs and scripts

nJim committed:
Merge branch '11.12.2-release' into SHS-6316-ddev

nJim committed:
feat(SHS-6316): reinstate lando as legacy option

nJim committed:
feat(SHS-6316): run npm commands within docker container

nJim committed:
fix(SHS-6316): fix for unique database names in ddev

nJim committed:
fix(SHS-6316): fix for unique database names in ddev

nJim committed:
fix(SHS-6316): fix for unique database names in ddev

nJim committed:
fix(SHS-6316): fix for unique database names in ddev

nJim committed:
fix(SHS-6316): fix for unique database names in ddev

nJim committed:
fix(SHS-6316): fix for unique database names in ddev

nJim committed:
fix(SHS-6316): fix for unique database names in ddev

nJim committed:
fix(SHS-6316): fix for unique database names in ddev

nJim committed:
Merge branch '11.13.1-release' into SHS-6316-ddev

nJim committed:
fix(SHS-6316): fix for unique database names in ddev

nJim committed:
fix(SHS-6316): fix for unique database names in ddev

nJim committed:
fix(SHS-6316): fix for unique database names in ddev

nJim committed:
Merge branch '11.14.1-release' into SHS-6316-ddev

nJim committed:
fix(SHS-6316): allow internal path to ddev sites

nJim committed:
Merge branch '11.15.1-release' into SHS-6316-ddev

nJim committed:
feat(SHS-6317): add codeception to ddev

nJim committed:
feat(SHS-6317): add codeception to ddev

nJim committed:
feat(SHS-6317): add codeception to ddev

ahughes3 commented:
[Tugboat](https://www.tugboatqa.com) has finished building the preview for this pull request!

Links:
- https://pr1966-e3e1xnjwaus8r4p6epvp8qxrc5dwfl3x.tugboatqa.com
- https://hs-colorful-e3e1xnjwaus8r4p6epvp8qxrc5dwfl3x.tugboatqa.com
- https://hs-traditional-e3e1xnjwaus8r4p6epvp8qxrc5dwfl3x.tugboatqa.com
- https://english-e3e1xnjwaus8r4p6epvp8qxrc5dwfl3x.tugboatqa.com
- https://history-e3e1xnjwaus8r4p6epvp8qxrc5dwfl3x.tugboatqa.com
- https://west-e3e1xnjwaus8r4p6epvp8qxrc5dwfl3x.tugboatqa.com


Dashboard:
- https://dashboard.tugboatqa.com/68c061ddf7f7fc63f7136d7c

Visual Diffs:
- https://dashboard.tugboatqa.com/68c061ddf7f7fc63f7136d7c/visualdiffs

cienvaras reviewed (changes requested):
@nJim I was able to run acceptance tests without any issue, but when I tried to run functional tests, I got errors like this for all tests:

```
 Test  codeception/functional/Install/Content/VideoEmbedCest.php:testVideoEmbed
                                                                                                                                                             
  [ConnectionException] Can't connect to WebDriver at http://127.0.0.1:4444/wd/hub. Make sure that ChromeDriver, GeckoDriver or Selenium Server is running.  
                                                                                                                                                             
#1  /var/www/html/vendor/symfony/event-dispatcher/EventDispatcher.php:220
#2  /var/www/html/vendor/symfony/event-dispatcher/EventDispatcher.php:56
#3  /var/www/html/vendor/bin/codecept:119
```

Is there anything I might be missing?

nJim committed:
feat(SHS-6317): add codeception to ddev

nJim committed:
feat(SHS-6317): add codeception to ddev

nJim commented:
@cienvaras Seems this fell into a 'works on my machine' type of problem. I has selenium manually setup but this was not repeatable. I've made three updates:
1. Added selenum information to the ddev config so that services can talk to each other
2. Increased the default timeout of tests
3. Added a theme compilation step for the testing site

This is again ready for testing. Follow this process to ensure a complete refresh:
```
# First destroy the containers:
ddev delete --omit-snapshot -y

# Then setup the test environment:
ddev setup-tests

# Finally run the testing and make sure everything work:
ddev blt codeception --group=install --suite=acceptance
ddev blt codeception --group=install --suite=functional
```

cienvaras reviewed (approved):
@nJim LGTM! Thanks!

@ahughes3 @Jnwibowo7321 Ready for you. This PR only impacts local development, so feel free to pass it straight to SWS.

joegl commented:
@cienvaras It looks like this was branched off of the #1906 PR which was just squashed-merged, so there are conflicts. Would you be able to resolve these?

cienvaras committed:
Merge branch '11.16.1-release' of github.com:SU-HSDO/suhumsci into SHS-6317-codeception

cienvaras commented:
@joegl Conflicts fixed

joegl reviewed (approved):
LGTM. I don't use DDEV and am happy to merge to enable this for DDEV users.
